generated/antimony.1
.ie \n(.g .ds Aq \(aq
.el .ds Aq '
.TH antimony 1  "antimony 5.2.1" 
.SH NAME
antimony \- Sandbox Applications
.SH SYNOPSIS
\fBantimony\fR [\fB\-h\fR|\fB\-\-help\fR] [\fB\-V\fR|\fB\-\-version\fR] <\fIsubcommands\fR>
.SH DESCRIPTION
Sandbox Applications
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version
.SH SUBCOMMANDS
.TP
antimony\-run(1)
Run a profile
.TP
antimony\-edit(1)
Edit an existing profile
.TP
antimony\-refresh(1)
Refresh caches
.TP
antimony\-integrate(1)
Integrate a profile into the user environment
.TP
antimony\-remove(1)
Remove features/profiles, or reset user definitions to the default
.TP
antimony\-export(1)
Export user profiles
.TP
antimony\-import(1)
Import user profiles
.TP
antimony\-info(1)
Get information about profiles/features
.TP
antimony\-package(1)
Package a profile into a self\-contained executable
.TP
antimony\-help(1)
Print this message or the help of the given subcommand(s)
.SH VERSION
v5.2.1


md/Changelog.md
## 5.0.0

### Breaking Changes

#### Antimony
* Profile/Feature libraries have been split into `files` and `directories` within a new `[libraries]` header. System profiles have been updated, but user profiles will need to be updated.
	* `libraries.no_sof` now controls whether roots are mounted, rather than simply having `/usr/lib` in the library list. The `lib` feature has been updated, so no action is needed.
	* `libraries.roots` now control per-profile library roots, which are added to a newly defined `library_roots` field in Antimony’s configuration. You will need to set your library roots if you are installing manually.
* Profile Hooks have been changed. Existing hooks will need to be migrated.
	* A mandatory `type` field now specifies between `Program`, `Shell`, and `Antimony` the first two are analogous to the old `content`/`path` scheme, with `Antimony` allowing privileged execution of another profile (Other hooks only run as the user). The `content` field now either accepts the script content, path, or profile name.
	* `args` has been renamed to `arguments` to align with the Profile field.
* Profile IPC has been renamed to use plural keys. User profiles may need to be updated:
	* `own` is now `owns`
	* `talk` in now `talks`
	* `call` is now `calls`
	* `see` is now `sees`
* The Configuration file has been moved to `/etc/antimony.toml`. Drop-in support is available by placing TOML files in `/etc/antimony.d` The old location is no longer supported; you will need to migrate your existing configuration. 
* The `config` command has been removed, as the configuration file is now a root-owned config in `/etc`.
* The `backend` command has been removed, and the database backend has been removed. There is no use case where it performs better than files. 
* The `cache_store` and `config_store` keys are no longer valid in the configuration file, and will need to be removed.
* Fedora (`.rpm`) packages are no longer built, and are no longer supported.

#### Utilities
* `antimony-bench` has been updated, and can only be used to benchmark version 2.4.0 and greater. 
* `antimony-dumper` and `antimony-monitor` now take an absolute path for the instance.

### New

* The configuration file now supports defining environment variables within the `[environment]` header. These values are only used if it is undefined, so setting `RUST_LOG = "info"` would set that value by default, but calling `RUST_LOG=trace antimony` would overwrite that, as would explicitly setting the value in the environment through `XDG_CONFIG_HOME/environment.d`, `.bashrc`, `/etc/profile`, etc.
* The `info` command has been brought back, dumping the TOML contents of profiles/feature, and diffing user and system definitions.
* The `integrate` command now allows you to create arbitrary profiles in `~/.local/bin` which lets you run programs within Antimony even if a profile is not defined (In the same way it works through the command line)
* The `edit` command now allows you to edit malformed profiles, though they must be valid in order for them to be saved.
* Antimony now ships with an AppArmor profile. You should no longer need to modify `sysctl` to get it, or any application run within it, to work.
* A front-end test suite has been added, allowing us to test `antimony` directly as a binary, rather than through Rust.

### Fixes

* Undoing integration via `antimony integrate -r` is now allowed even if the profile application has been uninstalled from the system.
* `antimony-lockdown` can no longer lose ownership of its home after an update
* `antimony-dumper` respects the `no-timeout` argument.
* `antimony-monitor` should no longer busy-wait as a zombie if the sandbox closes but Antimony does not signal it to terminate.
* `antimony-tracer` will no longer report paths in `/home/antimony`, such as in cases where you symlinked your home user to that path for localization.
### Profiles

### Features

* The `qt6` profile has been broken up into sub-features. User profiles may need to be updated.

### Crates

#### Common
* `common::singleton` can now have multiple instances.

#### Spawn
* `spawn` now performs `String` to `CString` on the `spawn()` call, so the `arg` suite of functions no longer return errors, and always succeed.
* `spawn` no longer has the `fork` feature. It was an interesting proof-of-concept, but had no utility.
* `Spawner` now supports a `env_or` function, that either takes the value from the environment, or falls back to the provided default.

### Technical

* `find` calls now filter on file type, and `get_dir` now checks for `*.so*` for non-executable libraries.
* `get_dir` now returns only the libraries that are external to that directory that are required, effectively treating the directory as it’s own library. 
* `proxy` is no longer initialized in its own thread.
* The `rayon` thread pool now initialized to half of hardware concurrency as it shows better performance on benchmarks.

generated/antimony.bash
_antimony() {
    local i cur prev opts cmd
    COMPREPLY=()
    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
        cur="$2"
    else
        cur="${COMP_WORDS[COMP_CWORD]}"
    fi
    prev="$3"
    cmd=""
    opts=""

    for i in "${COMP_WORDS[@]:0:COMP_CWORD}"
    do
        case "${cmd},${i}" in
            ",$1")
                cmd="antimony"
                ;;
            antimony,edit)
                cmd="antimony__subcmd__edit"
                ;;
            antimony,export)
                cmd="antimony__subcmd__export"
                ;;
            antimony,help)
                cmd="antimony__subcmd__help"
                ;;
            antimony,import)
                cmd="antimony__subcmd__import"
                ;;
            antimony,info)
                cmd="antimony__subcmd__info"
                ;;
            antimony,integrate)
                cmd="antimony__subcmd__integrate"
                ;;
            antimony,package)
                cmd="antimony__subcmd__package"
                ;;
            antimony,refresh)
                cmd="antimony__subcmd__refresh"
                ;;
            antimony,remove)
                cmd="antimony__subcmd__remove"
                ;;
            antimony,run)
                cmd="antimony__subcmd__run"
                ;;
            antimony__subcmd__help,edit)
                cmd="antimony__subcmd__help__subcmd__edit"
                ;;
            antimony__subcmd__help,export)
                cmd="antimony__subcmd__help__subcmd__export"
                ;;
            antimony__subcmd__help,help)
                cmd="antimony__subcmd__help__subcmd__help"
                ;;
            antimony__subcmd__help,import)
                cmd="antimony__subcmd__help__subcmd__import"
                ;;
            antimony__subcmd__help,info)
                cmd="antimony__subcmd__help__subcmd__info"
                ;;
            antimony__subcmd__help,integrate)
                cmd="antimony__subcmd__help__subcmd__integrate"
                ;;
            antimony__subcmd__help,package)
                cmd="antimony__subcmd__help__subcmd__package"
                ;;
            antimony__subcmd__help,refresh)
                cmd="antimony__subcmd__help__subcmd__refresh"
                ;;
            antimony__subcmd__help,remove)
                cmd="antimony__subcmd__help__subcmd__remove"
                ;;
            antimony__subcmd__help,run)
                cmd="antimony__subcmd__help__subcmd__run"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        antimony)
            opts="-h -V --help --version run edit refresh integrate remove export import info package help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__edit)
            opts="-f -s -h --feature --system --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__export)
            opts="-n -d -f -s -h --name --dest --feature --system --seccomp --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --name)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -n)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --dest)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                -d)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help)
            opts="run edit refresh integrate remove export import info package help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__edit)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__export)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__help)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__import)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__info)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__integrate)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__package)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__refresh)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__remove)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__help__subcmd__run)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__import)
            opts="-f -s -h --feature --system --seccomp --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__info)
            opts="-f -s -h --feature --diff --system --seccomp --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__integrate)
            opts="-r -s -a -e -o -h --remove --shadow --autostart --enable --config-mode --create-desktop --overwrite --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --config-mode)
                    COMPREPLY=($(compgen -W "action file" -- "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__package)
            opts="-d -v -h --dest --version --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --dest)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                -d)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                --version)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -v)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__refresh)
            opts="-d -s -h --dry --hard --seccomp --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__remove)
            opts="-f -y -h --feature --yes --seccomp --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        antimony__subcmd__run)
            opts="-l -c -h --dry --refresh --path --dir --lockdown --config --features --conflicts --inherits --home-policy --home-name --home-path --home-lock --home-lock-policy --seccomp --portals --sees --talks --owns --calls --disable-ipc --system-bus --user-bus --harden-ipc --file-passthrough --ro --rw --link --temp --binaries --libraries --directories --roots --no-sof --devices --namespaces --env --preserve-env --sandbox-args --ignored-exit-codes --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --path)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --dir)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                --lockdown)
                    COMPREPLY=($(compgen -W "true false" -- "${cur}"))
                    return 0
                    ;;
                -l)
                    COMPREPLY=($(compgen -W "true false" -- "${cur}"))
                    return 0
                    ;;
                --config)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -c)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --features)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --conflicts)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --inherits)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --home-policy)
                    COMPREPLY=($(compgen -W "none enabled read-only overlay" -- "${cur}"))
                    return 0
                    ;;
                --home-name)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --home-path)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                --home-lock)
                    COMPREPLY=($(compgen -W "true false" -- "${cur}"))
                    return 0
                    ;;
                --home-lock-policy)
                    COMPREPLY=($(compgen -W "notify abort overlay" -- "${cur}"))
                    return 0
                    ;;
                --seccomp)
                    COMPREPLY=($(compgen -W "disabled permissive enforcing notifying" -- "${cur}"))
                    return 0
                    ;;
                --portals)
                    COMPREPLY=($(compgen -W "background camera clipboard documents file-chooser global-shortcuts inhibit location notifications open-uri proxy-resolver realtime screen-cast screenshot settings secret network-monitor remote-desktop input-capture" -- "${cur}"))
                    return 0
                    ;;
                --sees)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --talks)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --owns)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --calls)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-passthrough)
                    COMPREPLY=($(compgen -W "read-only read-write executable" -- "${cur}"))
                    return 0
                    ;;
                --ro)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --rw)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --link)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --temp)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --binaries)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --libraries)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --directories)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                --roots)
                    COMPREPLY=()
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o plusdirs
                    fi
                    return 0
                    ;;
                --devices)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --namespaces)
                    COMPREPLY=($(compgen -W "all user ipc pid net uts c-group" -- "${cur}"))
                    return 0
                    ;;
                --env)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --preserve-env)
                    COMPREPLY=($(compgen -W "true false" -- "${cur}"))
                    return 0
                    ;;
                --sandbox-args)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --ignored-exit-codes)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

if [[ "${BASH_VERSINFO[0]}" -eq 4 && "${BASH_VERSINFO[1]}" -ge 4 || "${BASH_VERSINFO[0]}" -gt 4 ]]; then
    complete -F _antimony -o nosort -o bashdefault -o default antimony
else
    complete -F _antimony -o bashdefault -o default antimony
fi
